share/shwap/p2p/shrex/client.go

```diff
@@ -0,0 +1,184 @@
+package shrex
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"go.uber.org/zap"
+
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	shrexpb "github.com/celestiaorg/celestia-node/share/shwap/p2p/shrex/pb"
+)
+
+// Client implements client side of shrex protocol to obtain data from remote
+// peers.
+type Client struct {
+	params *ClientParams
+
+	host    host.Host
+	metrics *Metrics
+}
+
+// NewClient creates a new shrEx client
+func NewClient(params *ClientParams, host host.Host) (*Client, error) {
+	if err := params.Validate(); err != nil {
+		return nil, fmt.Errorf("shrex/client: parameters are not valid: %w", err)
+	}
+	return &Client{
+		host:   host,
+		params: params,
+	}, nil
+}
+
+func (c *Client) WithMetrics() error {
+	metrics, err := InitClientMetrics()
+	if err != nil {
+		return fmt.Errorf("shrex/client: init Metrics: %w", err)
+	}
+	c.metrics = metrics
+	return nil
+}
+
+func (c *Client) Get(
+	ctx context.Context,
+	req request,
+	resp response,
+	peer peer.ID,
+) error {
+	logger := log.With(
+		"source", "client",
+		"name", req.Name(),
+		"peer", peer.String(),
+	)
```

Comment:
Can we add fileds from request here? Otherwise it is hard to understand which request has failed

Reply:
Only the height field is left. Will add it

Reply:
request is an interface w/o access to these fields. These fields can be added in getter